blob/parser.go

```diff
@@ -0,0 +1,151 @@
+package blob
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/pkg/shares"
+)
+
+// parser is a helper struct that allows collecting shares and transforming them into the blob.
+// it contains all necessary information that is needed to build the blob:
+// * position of the blob inside the EDS;
+// * blob's length;
+// * shares needed to build the blob;
+// * extra condition to verify the final blob.
+type parser struct {
+	index    int
+	length   int
+	shares   []shares.Share
+	verifyFn func(blob *Blob) bool
+}
+
+// skipPadding skips first share in the range if this share is the Padding share.
+func (p *parser) skipPadding(shares []shares.Share) ([]shares.Share, error) {
+	if len(shares) == 0 {
+		return nil, errEmptyShares
+	}
+
+	isPadding, err := shares[0].IsPadding()
+	if err != nil {
+		return nil, err
+	}
+
+	if !isPadding {
+		return shares, nil
+	}
+
+	// update blob index if we are going to skip one share
+	p.index++
+	if len(shares) > 1 {
+		return shares[1:], nil
+	}
+	return nil, nil
+}
+
+// NOTE: passing shares here needed to detect padding shares(as we do not need this check in addShares)
+func (p *parser) set(index int, shrs []shares.Share) ([]shares.Share, error) {
+	if len(shrs) == 0 {
+		return nil, errEmptyShares
+	}
+
+	shrs, err := p.skipPadding(shrs)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(shrs) == 0 {
+		return nil, errEmptyShares
+	}
+
+	// `+=` as index could be updated in `skipPadding`
+	p.index += index
+	length, err := shrs[0].SequenceLen()
+	if err != nil {
+		return nil, err
+	}
+
+	p.length = shares.SparseSharesNeeded(length)
+	return shrs, nil
+}
+
+// addShares sets shares until the blob is completed and returns extra shares back.
+// we do not need here extra condition to check padding shares as we do not expect it here.
+// it is possible only between two blobs.
+func (p *parser) addShares(shares []shares.Share) (shrs []shares.Share, isComplete bool) {
+	index := -1
+	for i, sh := range shares {
+		p.shares = append(p.shares, sh)
+		if len(p.shares) == p.length {
+			index = i
+			isComplete = true
+			break
+		}
+	}
+
+	if index == -1 {
+		return
+	}
+
+	if index+1 >= len(shares) {
+		return shrs, true
+	}
+	return shares[index+1:], true
+}
+
+// transform parses shares and creates the Blob.
+func (p *parser) transform() (*Blob, error) {
```

Comment:
Last nit. If this is a parser and transform is its main operation, it makes sense to name the method as `parse`.

Also consider rearranging methods order for readabilitiy

Reply:
makes sense